update readme and documentation (#1124)
USRSE/usrse.github.io: docs/tests_ci.md

needing to be connected to CircleCI, and under settings:

 - build forked pull requests should be on
 - cancel redundant builds is suggested
 - workflows should be enabled

If you want to edit any of the tests, you should edit [config.yml](.circleci/config.yml).
Details about running tests locally are included below. This can be good to do if you
change an input file in [_data](_data) and want to test it.

### Previewing the Site 

CircleCI builds the site and creates preview of individual pages. 
The preview only works for individual pages (the links within the preview site don't link to other pages in preview).  

In the PR, you'll see:

![PR checks notification](assets/img/pr_checks.png)

Click on Show All Checks, then get Details for the CircleCI checks:

![PR checks details](assets/img/pr_checks_detailed.png)

This will take you to the CircleCI page, where you should click on the Artifacts link:

![CircleCI project page](assets/img/circleci.png)

This will open a list of the individual pages - click on links to review the pages (reminder: internal links won't go to other preview pages)

![CircleCI preview links](assets/img/circleci_artifacts.png)

> ### Test Jobs

Reviewer: "Test jobs" is a bit confusing and hard to parse. This could be testing of job additions, or jobs that test the website. I wonder if another word would help, perhaps "testing additions of new jobs"?

Reviewer 2: I agree with @danielskatz

Jobs are tested for correctness, meaning that all fields are entered, a date string

> is entered for the "expires" field, and the url is valid. You can run tests locally 

Reviewer 2: ```suggestion
is entered for the "expires" field, and the URL is valid. You can run tests locally 
```

like:

```bash
$ cd tests
$ python -m unittest test_jobs
```

> ### Count Jobs

Reviewer: maybe add something about what this is used for at the start of the description?

Author: I don't actually know - this is just content I moved over.  Anyone else know?

Reviewer: I think it's used to count the number of jobs that have been added (for the annual meeting, etc.)

A [script](scripts/count_jobs.py) is provided that will clone the repository
to a temporary directory, find all commits with a changed job file,
and then checkout and read each commit to get the jobs present for that time.

> We then use the title and url for the job as a unique identifier to determine

Reviewer 2: ```suggestion
We then use the title and URL for the job as a unique identifier to determine
```

> if the job has been seen. A job with the same name and url, and thus the same

Reviewer 2: ```suggestion
if the job has been seen. A job with the same name and URL, and thus the same
```

unique identifier, is considered the same job. You can run this script
as is if you just want to derive counts:

```bash
$ python scripts/count_jobs.py
Cloning repository https://github.com/USRSE/usrse.github.io
Found 43 commits for _data/jobs.yml
Found a total of 35 unique jobs across 43 commits.
```

or you can add an output file to save the compiled job content to file

```bash
$ python scripts/count_jobs.py all-jobs.yml
Cloning repository https://github.com/USRSE/usrse.github.io
Found 43 commits for _data/jobs.yml
Found a total of 35 unique jobs across 43 commits.
Saving to output file /home/vanessa/Desktop/Code/usrse/usrse.github.io/all-jobs.yml
```

The repository is always cleaned up, and the parsing done separately from the
script.

### GitHub CI

> #### URLChecker and Spelling

Reviewer: URLChecker, URLschecker, or URLs-checker?

> The [URLschecker](https://github.com/urlstechie/URLs-checker) is a GitHub action

Reviewer 2: ```suggestion
The [urlchecker](https://github.com/urlstechie/urlchecker-action) is a GitHub action
```

that @vsoch worked on to contribute retry and some other nice features for the 
repository here. These features are available as of version 0.1.6 that is used
in the [workflow](.github/workflows/urlchecker.yml). Relative URLs internal to the site are not checked.  

If there are URLs that should be systematically ignored by the checker, they can be added to .github/workflows/linting.yaml 

In addition, @vsoch found
a Rust tool called [crate-ci/typos](https://github.com/marketplace/actions/typos-action)

> and contributed an equivalent action so all posts and pages are spell checked.

Reviewer 3: ```suggestion
and contributed an equivalent action so all posts and pages are spell-checked.
```

If your CI fails, the spelling suggestions will be shown and you can manually
update the mistakes, or [install typos](https://github.com/crate-ci/typos#install) 
and have all errors fixed automatically:

```bash
typos ./pages ./_posts ./README.md --write-changes
```

The config file to specify words to ignore is 

> [`.github/workflows/typo_config.toml`](.github/workflows/topy_config.toml).

Reviewer 3: ```suggestion
[`.github/workflows/typo_config.toml`](.github/workflows/typo_config.toml).
```
:smile:

Edit this file if there are phrases or words that need to be ignored (e.g., surnames, acronyms).

#### Clean Expired Jobs

The workflow [clean-expired-jobs.yml](.github/workflows/clean-expired-jobs.yml) is run nightly,
and uses the same function from the urlchecker to check for expired links in jobs.yml,

> and given an expired link, remove it from the file if the url check fails. In the case

Reviewer 2: ```suggestion
and given an expired link, remove it from the file if the URL check fails. In the case
```

that a link is not expired and the check fails, we would want to know about this
(and the test will fail). For all jobs, we don't remove them immediately upon expiration -
we give the submitter 60 days to possibly update the data file with a later expiration date.

This job needs to be updated in conjunction with the URL checker. If they are not on
compatible URL checker versions, you may receive inconsistent behavior and erroneous
failures between this and the PR linting job.

#### Post New Jobs to Slack

The workflow [jobs-poster.yaml](.github/workflows/jobs-poster.yaml) is run on any push
to `main` with changes to `_data/jobs.yml`. If new jobs are found, it will post the Job URL to
the USRSE Slack `#jobs` channel. It utilizes the [Jobs updater](https://github.com/rseng/jobs-updater)

> Github Action by @vsoch and @jhkennedy to parse the `_data/jobs.yml` file for new jobs and post them

Reviewer 3: ```suggestion
GitHub Action by @vsoch and @jhkennedy to parse the `_data/jobs.yml` file for new jobs and post them
```

the USRSE Slack. For the action:

>  - unique: determines the field in the jobs.yml that determines uniqueness (defaults to url)

Reviewer 2: ```suggestion
 - `unique`: determines the field in the `jobs.yml` that determines uniqueness (defaults to URL)
```

>  - keys: a comma separated list of fields to include. All except for url will have a prefix, so it's recommended to put the url last.

Reviewer 2: ```suggestion
 - `keys`: a comma-separated list of fields to include. All except for URL will have a prefix, so it's recommended to put the URL last.
```

The other fields are intuitive. 

> Example output (in the console that might go to Slack or Twitter)

Reviewer: maybe say that we would love someone to make this work for Mastodon too?

you should sign into the [developer portal](https://developer.twitter.com/en/portal/) and make a new project
that describes the goal of US-RSE, e.g.,:

> We (the United States Research Software Engineer Association) use Twitter to get a broad reach to people working on software in academia - to support community spirit and growth. We currently have a jobs board on our website that populates from GitHub, and will reach more potentially interested people in our community by posting new jobs (as they are merged in a workflow) from a GitHub workflow.

For the use case, you can choose "making a bot" and the title for the project can be anything you like.
Importantly, once you create the bot you'll need to add the following secrets to your GitHub repository:

 - TWITTER_ACCESS_TOKEN: is the key/token created for the user account
 - TWITTER_ACCESS_SECRET: is the secret created for the user account
 - TWITTER_CONSUMER_API_KEY: is the main key for the developer app
 - TWITTER_CONSUMER_API_SECRET: is main secret for the developer app
 
Yes, this means that the tokens are specific to this account.

#### Greetings

> This simple greetings action greets first time users (for issues).

Reviewer 3: ```suggestion
This simple greetings action greets first-time users (for issues).
```

The logic of this is determined by the [greetings.yml](.github/workflows/greetings.yml)
workflow. 

#### Member Counts

Two scripts help to create a branch with an updated [member counts file](_data/memberCounts.csv)
that starts with the prefix `update/member-counts`. The workflow [member-counts.yaml](.github/workflows/member-counts.yaml) will generate an updated file and commit and push to a new branch, and it uses [pull-request.sh](scripts/pull-request.sh) to then open a PR with the new branch to the repository. For GitHub CI, there are currently no secrets or credentials, and no setup is required - having actions enabled for the repository and placing the file under `.github/workflows`
enables it.

## Rakefile

A legacy [Rakefile](Rakefile) is kept with the repository but it is no longer used.

## Frequently Asked Questions

> Why do we use different services?

Using multiple "free tier" CI services is a common thing for open source projects to do.
There are several reasons to do this:

>  1. we can better leverage a free tier, meaning a maximum number of jobs run in parallel or minutes per month by spreading work over multiple services. 
>  2. we can scope a particular kind of test to a service. For example, one service might just be to test the core software, another might be to build and deploy containers, and a third might be to preview a site.
>  3. each CI service offers unique features. For example, GitHub has the closest integration with the repository here, and CircleCI allows us to preview artifacts.

Reviewer 3: ```suggestion
 1. We can better leverage a free tier, meaning a maximum number of jobs run in parallel or minutes per month by spreading work over multiple services. 
 2. We can scope a particular kind of test to a service. For example, one service might just be to test the core software, another might be to build and deploy containers, and a third might be to preview a site.
 3. Each CI service offers unique features. For example, GitHub has the closest integration with the repository here, and CircleCI allows us to preview artifacts.
```
